### PR TITLE
feat(native): shared audit infra and doc-only rule resolutions (#247)

### LINKED ISSUE #247: test(native): 9 doc-only rules have no implementation — need implementation or intentional stub decision
## Summary

The auto-generated [Rule Catalog](docs/rules/RULE_CATALOG.md) identifies 9 rules that have `.md` documentation files but **no corresponding implementation** (no `*_graph.py` Python file in native rules, no Rego policy, no Ansible validator code).

These rules need either:
- A concrete implementation (GraphRule subclass or equivalent)
- An intentional decision to mark them as planned/deferred stubs

## Affected rules

| Rule | Description | Notes |
|------|-------------|-------|
| **L031** | File permission may be insecure (annotation-based) | May overlap with OPA L020/L021 mode checks |
| **M029** | Inventory scripts must include `_meta.hostvars` in JSON output (2.23) | Also listed in #231 as missing doc examples |
| **P001** | Validate module name (Ansible required) | May be emitted by Ansible validator argspec checks (L058/L059) |
| **P002** | Validate module argument keys (Ansible required) | Same — may be covered by argspec validation |
| **P003** | Validate module argument values (Ansible required) | Same |
| **P004** | Validate variables (Ansible required) | Same |
| **R402** | Report variables used at end of sequence | Informational/reporting rule |
| **R404** | Expose variable_set for the task | Informational/reporting rule |
| **R501** | Suggest collection/role dependency | Advisory/suggestion rule |

## Questions to resolve

1. **P001–P004**: Are these intentionally documented as the Ansible validator's policy checks (emitted by L058/L059 argspec validation)? If so, they may need to be mapped to the Ansible validator in the catalog rather than Native.
2. **R402, R404, R501**: Are these planned features or aspirational stubs? They appear to be informational/advisory rules that would require deeper graph analysis.
3. **L031**: Does OPA L020/L021 already cover file permission checks sufficiently?
4. **M029**: Needs both implementation and doc examples (#231).

## Discovery

Generated by `tools/generate_rule_catalog.py` — the "Coverage Gaps → Doc-only rules" section.

## Related

- #231 — native rules missing YAML doc examples (M029 overlaps)

### PR DESCRIPTION
**Pull request by djdanielsson**
## Summary

Split from #347 (per review). This is **PR 1 of 3** — shared infrastructure and doc-only rule resolutions without R402/R404 implementations.

- Adds sensitivity/audit-metadata plumbing, graph-rule opt-in loading, and gateway persistence for audit metadata
- Resolves remaining doc-only rule gaps: P001–P004 delegated to Ansible validator, L031/M029 stubbed, R501 planned
- Refactors L039/L110 to shared `_variable_helpers.py`; consolidates violation conversion into `daemon/violation_convert.py`

Fixes #247 (partial — R402/R404 follow in stacked PRs)

## Changes

- `sensitivity.py` / `audit_metadata.py`: redaction helpers for audit payloads (`variables_used`, `variable_set`, `inbound_src`)
- `validate.proto`: `graph_rule_opt_in` field for disabled-by-default R402/R404
- Gateway: `audit_metadata` column + REST exposure with re-sanitization on read
- `tools/generate_rule_catalog.py`: improved test-detection heuristics and doc-only rule status
- Review follow-ups: non-JSON audit string sanitization, malformed JSON logging, URL userinfo redaction edge case

## Stacked PRs

| PR | Branch | Scope |
|----|--------|-------|
| **This PR** | `fix/247-doc-only-rule-infra` | Shared infra + doc-only resolutions |
| Next | `feat/R402-list-all-used-variables` | R402 rule (base: this branch) |
| Next | `feat/R404-show-variables` | R404 rule (base: this branch) |

## Test plan

- [x] `tox -e unit`
- [x] `tox -e lint` (pre-commit on commit)
- [x] CI green on latest push (`1bb6987b`)

## Follow-up issues

- #373 — catalog tested heuristic for planned rules
- #375 — register delegated P001–P004 in Gateway catalog
- #376 — bump grpcio floor to match codegen guard

## Summary by CodeRabbit

* **New Features**
  * Added per-scan GraphRule opt-in support for disabled-by-default rules during validation and remediation.
  * Exposed `audit_metadata` in violation details, including API/OpenAPI support, with sanitization for safe display and storage.
* **Bug Fixes**
  * Improved audit-metadata JSON parsing with graceful handling of malformed/non-serializable values.
  * Strengthened scan/rescan behavior for aggregate-scope correctness and play-scoped remediation.
* **Documentation**
  * Regenerated the rule catalog coverage and updated the API schema for `audit_metadata`.
* **Tests**
  * Expanded coverage for opt-in rule loading and audit-metadata conversion/redaction end-to-end.

### CODE REVIEW

**Comment by djdanielsson** [comment]
Follow-up items from contributor review — tracked as issues:

- #373 — tighten catalog **Tested** heuristic so R402/R404 aren't marked tested from infra-only references
- #374 — fix `_variable_helpers.py` docstring drift (sensitivity helpers import location)
- #375 — register delegated P001–P004 in ADR-041 Gateway catalog (`collect_all_rules`)

Related open issues: #247 (doc-only rules), #248 (broader test coverage)

**Comment by djdanielsson** [comment]
@coderabbitai review

**Comment by djdanielsson** [comment]
@coderabbitai full review

**Comment by djdanielsson** [comment]
@coderabbit full review

File: src/apme_engine/daemon/violation_convert.py [hunk 1]
```diff
@@ -217,7 +228,30 @@ def violation_dict_to_proto(v: ViolationDict | Mapping[str, str | int | list[int
...
+                        if serialized is not None:
+                            out.metadata[key] = serialized
+                else:
+                    serialized = serialize_audit_metadata_value(
+                        val,
+                        rule_id=str(v.get("rule_id") or ""),
+                        key=key,
+                    )
+                    if serialized is not None:
+                        out.metadata[key] = serialized
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/daemon/violation_convert.py:255]: Good catch — the `json.JSONDecodeError` branch now runs non-JSON audit strings through `sanitize_audit_metadata_value` + `serialize_audit_metadata_value` instead of storing the raw value. Added tests for sensitive plain strings (`variable_set`) and URL credentials (`inbound_src`). Fixed in b981db6.

File: src/apme_gateway/api/router.py [hunk 2]
```diff
@@ -1411,6 +1413,15 @@ def _to_violation_detail(
...
         suppressed = fp_rule in rule_only_hashes
+    audit_metadata: dict[str, object] | None = None
+    raw_audit = getattr(v, "audit_metadata", "") or ""
+    if raw_audit:
+        try:
+            parsed_audit = json.loads(raw_audit)
+            if isinstance(parsed_audit, dict):
+                audit_metadata = {key: sanitize_audit_metadata_value(key, value) for key, value in parsed_audit.items()}
+        except json.JSONDecodeError:
+            audit_metadata = None
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_gateway/api/router.py:1484]: Added a `logger.warning` on `json.JSONDecodeError` when decoding persisted `audit_metadata`, including violation id and rule_id for correlation. Added `test_get_scan_detail_logs_malformed_audit_metadata` to assert the warning and `audit_metadata=None` response. Fixed in b981db6.

File: src/apme/v1/ansible_pb2_grpc.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from apme.v1 import common_pb2 as apme_dot_v1_dot_common__pb2
 from apme.v1 import validate_pb2 as apme_dot_v1_dot_validate__pb2
 
-GRPC_GENERATED_VERSION = '1.78.0'
+GRPC_GENERATED_VERSION = '1.81.1'
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/ansible_pb2_grpc.py]: Valid finding — the PR had accidentally regenerated stubs with `GRPC_GENERATED_VERSION = 1.81.1` while `uv.lock` still pins `grpcio==1.80.0`, which would fail the import guard at runtime. Reverted the unrelated `*_pb2_grpc.py` version bumps in b981db6 (only `validate.proto` changed in this PR). Tracked the proper project-wide grpcio floor bump separately in #376.

File: src/apme/v1/common_pb2_grpc.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 import warnings
 
 
-GRPC_GENERATED_VERSION = '1.78.0'
+GRPC_GENERATED_VERSION = '1.81.1'
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/common_pb2_grpc.py]: Same root cause as the ansible stub comment — accidental `GRPC_GENERATED_VERSION` bump vs locked `grpcio==1.80.0`. Reverted generated stub version lines to match `main` in b981db6; follow-up dependency alignment tracked in #376.

**Comment by djdanielsson** [comment]
@CodeRabbit full review

File: src/apme_engine/engine/sensitivity.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    Returns:
+        URL with userinfo replaced by ``[REDACTED]``.
+    """
+    match = _URL_USERINFO_RE.match(url)
+    if not match:
+        return url
+    scheme = match.group(1)
+    rest = url[match.end() :]
+    return f"{scheme}://{REDACTED}@{rest}"
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/engine/sensitivity.py:70]: Valid finding — the old regex split user/password at the first `@`, so a password like `p@ss` leaked `ss` into the redacted URL. Updated `_URL_USERINFO_RE` to match the full userinfo segment (`scheme://...userinfo@`) before the host and added `test_redact_url_userinfo_password_with_at_sign`. Fixed in d96200b.

**Comment by djdanielsson** [comment]
@CodeRabbit full review

**Comment by djdanielsson** [comment]
@CodeRabbit full review

File: docs/rules/RULE_CATALOG.md [hunk 6]
```diff
@@ -352,54 +352,20 @@
...
-- **M030** (Native): Conditional expressions that fail Jinja2 parsing will error in 2.23 instead of being silently ignored
-- **R105** (Native): Outbound transfer (annotation-based).
-- **R106** (Native): Inbound transfer (annotation-based).
-- **R107** (Native): Package install with insecure option (annotation-based).
-- **R109** (Native): Key/config change (annotation-based).
-- **R115** (Native): File deletion (annotation-based).
-- **R117** (Native): Role is from Galaxy/external source.
-- **R401** (Native): Report inbound transfer sources.
+- **L031** (Native): [Stub] File permission may be insecure (annotation-based). — Covered by OPA rules L020/L021 which check file mode arguments directly. This annotation-based variant is retained as a placeholder for future annotation-driven detection (requires engine annotation pipeline).
+- **M029** (Native): [Stub] Inventory scripts must include _meta.hostvars in JSON output (enforced in 2.23) — Detection requires executing the inventory script at runtime to inspect its JSON output. Static analysis cannot determine _meta presence. Disabled by design until a runtime-analysis approach is approved.
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply docs/rules/RULE_CATALOG.md]: Wrapped `_meta.hostvars` and `_meta` in backticks in `M029_inventory_script_missing__meta.md` frontmatter (`description` + `status_reason`) so the generated `RULE_CATALOG.md` coverage-gaps entry no longer triggers MD037 emphasis lint. Regenerated catalog via the rule-catalog hook. Fixed in 7455b7c.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme/v1/common_pb2.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # NO CHECKED-IN PROTOBUF GENCODE
 # source: apme/v1/common.proto
-# Protobuf Python Version: 7.35.0
+# Protobuf Python Version: 6.33.5
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/common_pb2.py]: Regenerated all checked-in protobuf modules with protobuf 7.35.1 (runtime guards now validate 7.35.1). Fixed in 2ab8ad51.

File: src/apme/v1/ansible_pb2.py [hunk 8]
```diff
@@ -11,9 +11,9 @@
 from google.protobuf.internal import builder as _builder
 _runtime_version.ValidateProtobufRuntimeVersion(
     _runtime_version.Domain.PUBLIC,
-    7,
-    35,
-    0,
+    6,
+    33,
+    5,
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/ansible_pb2.py]: Regenerated ansible_pb2.py with protobuf 7.35.1 so generator metadata and ValidateProtobufRuntimeVersion match pyproject.toml. Fixed in 2ab8ad51.

File: src/apme/v1/primary_pb2_grpc.py [hunk 9]
```diff
@@ -6,7 +6,7 @@
 from apme.v1 import common_pb2 as apme_dot_v1_dot_common__pb2
 from apme.v1 import primary_pb2 as apme_dot_v1_dot_primary__pb2
 
-GRPC_GENERATED_VERSION = '1.82.1'
+GRPC_GENERATED_VERSION = '1.81.1'
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/primary_pb2_grpc.py]: Regenerated gRPC stubs with grpcio-tools 1.83.0; GRPC_GENERATED_VERSION now matches the declared grpcio>=1.82.1 floor. Fixed in 2ab8ad51.

File: src/apme/v1/validate_pb2.py [hunk 10]
```diff
@@ -2,7 +2,7 @@
 # Generated by the protocol buffer compiler.  DO NOT EDIT!
 # NO CHECKED-IN PROTOBUF GENCODE
 # source: apme/v1/validate.proto
-# Protobuf Python Version: 7.35.0
+# Protobuf Python Version: 6.33.5
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/validate_pb2.py]: Regenerated validate_pb2.py with protobuf 7.35.1 alongside the other proto modules. Fixed in 2ab8ad51.

File: tests/test_generate_rule_catalog.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+        def fake_import(
+            name: str,
+            globals: Mapping[str, object] | None = None,
+            locals: Mapping[str, object] | None = None,
+            fromlist: Sequence[str] = (),
+            level: int = 0,
+        ) -> object:
+            if name == "yaml":
+                raise ImportError("PyYAML not installed")
+            return real_import(name, globals, locals, fromlist, level)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_generate_rule_catalog.py:257]: Renamed fake_import parameters to globals_ and locals_ to avoid shadowing builtins (Ruff A002). Fixed in 2ab8ad51.

File: tools/generate_rule_catalog.py [hunk 12]
```diff
@@ -178,8 +230,19 @@ def _get_test_cache() -> dict[str, list[str]]:
...
+        seen_ids: set[str] = set()
         for m in all_rule_id_pat.finditer(text):
             rid = m.group(1)
+            if _is_incidental_reference(rid, text, py.name):
+                continue
+            seen_ids.add(rid)
+        for m in import_rule_id_pat.finditer(text):
+            rid = (m.group(1) or m.group(2) or (m.group(3) or "").upper()).strip()
+            if rid:
+                seen_ids.add(rid)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tools/generate_rule_catalog.py:252]: Removed import-only and grouped-module import heuristics from _get_test_cache; test coverage now requires quoted rule_id references that pass _is_incidental_reference. Updated tests accordingly. Fixed in 2ab8ad51.

File: tools/generate_rule_catalog.py [hunk 13]
```diff
@@ -120,32 +140,59 @@ def _get_fixable_ids() -> set[str]:
...
+        rf"(?:class\s+\w*{re.escape(rule_id)}\w*|"
+        rf"def\s+test_\w*{re.escape(rid_lower)}\w*|"
+        rf"import\s+.*{re.escape(rule_id)}|"
+        rf"Rule.*{re.escape(rule_id)})",
+        re.IGNORECASE,
+    )
+    if rule_context.search(text):
+        return False
+    rule_id_assignment = re.compile(rf'rule_id\s*[=:]\s*["\']{re.escape(rule_id)}["\']')
+    return not rule_id_assignment.search(text)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tools/generate_rule_catalog.py:182]: Updated _is_incidental_reference to accept quoted rule_id mapping keys in the assignment regex. Fixed in 2ab8ad51.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/audit_metadata.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+AUDIT_JSON_METADATA_KEYS: frozenset[str] = frozenset({"variables_used", "variable_set", "inbound_src"})
+
+
+def _sanitize_scalar(key: str, value: object) -> object:
+    if key == "inbound_src" and isinstance(value, str):
+        return redact_url_userinfo(value)
+    if isinstance(value, str) and value_looks_sensitive(value):
+        return REDACTED
+    return value
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/audit_metadata.py:26]: Extended inbound_src sanitization to recurse into dict/list values via redact_sensitive_structure, and applied the same pattern in R401_list_all_inbound_src_graph.py. Added regression test. Fixed in eeefe435.

File: src/apme_engine/remediation/graph_engine.py [hunk 15]
```diff
@@ -733,6 +734,7 @@ async def _rescan_and_record(
             Fresh violations from the rescan.
         """
         dirty = graph.dirty_nodes
+        effective_dirty = expand_dirty_node_ids(graph, self._rules, dirty)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/remediation/graph_engine.py:796]: expand_dirty_node_ids now receives enabled-only rules, matching rescan_dirty. Fixed in eeefe435.

File: src/apme/v1/validate_pb2_grpc.py [hunk 16]
```diff
@@ -6,7 +6,7 @@
 from apme.v1 import common_pb2 as apme_dot_v1_dot_common__pb2
 from apme.v1 import validate_pb2 as apme_dot_v1_dot_validate__pb2
 
-GRPC_GENERATED_VERSION = '1.82.1'
+GRPC_GENERATED_VERSION = '1.83.0'
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme/v1/validate_pb2_grpc.py:9]: Regenerated gRPC stubs with grpcio-tools 1.82.1 to match uv.lock; GRPC_GENERATED_VERSION is now 1.82.1. Fixed in eeefe435.

File: tests/test_violation_convert.py [hunk 17]
```diff
@@ -187,6 +187,88 @@ def test_missing_original_yaml_defaults_empty(self) -> None:
...
+        """Non-JSON inbound_src strings scrub embedded URL credentials."""
+        v: ViolationDict = {
+            "rule_id": "R401",
+            "inbound_src": "https://admin:pass@host/file.tar.gz",
+        }
+        proto = violation_dict_to_proto(v)
+        stored = proto.metadata["inbound_src"]
+        assert "[REDACTED]" in stored
+        assert "pass" not in stored
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_violation_convert.py]: Added assert admin not in stored so both URL credential components must be redacted. Fixed in eeefe435.

File: tools/generate_rule_catalog.py [hunk 18]
```diff
@@ -120,32 +140,61 @@ def _get_fixable_ids() -> set[str]:
...
+        rf"import\s+.*{re.escape(rule_id)}|"
+        rf"Rule.*{re.escape(rule_id)})",
+        re.IGNORECASE,
+    )
+    if rule_context.search(text):
+        return False
+    rule_id_assignment = re.compile(
+        rf'["\']?rule_id["\']?\s*[=:]\s*["\']{re.escape(rule_id)}["\']',
+    )
+    return not rule_id_assignment.search(text)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tools/generate_rule_catalog.py:182]: Removed import-only context from _is_incidental_reference, tightened Rule pattern to word boundary, and regenerated RULE_CATALOG.md (tested count 105/156). R402/R404 still show tested due to rule_id assignments in violation-convert tests; broader heuristic tracked in #373. Fixed in eeefe435.

**Comment by djdanielsson** [comment]
**Outside-diff review (graph_engine.py:736-757) — validated and fixed in eeefe435**

CodeRabbit flagged that `_resolve_dirty_violations` uses `effective_dirty` but `_rescan_fn` received only the original `dirty` set. **This is accurate**: when the gRPC rescan bridge runs without native (or before native returns), expanded aggregate ancestors could be marked fixed without being rescanned.

**Fix:** `_rescan_and_record` now passes `effective_dirty` to `_rescan_fn`, so the bridge forwards the full expanded set to native via `dirty_node_ids`. The built-in `rescan_dirty` fallback still receives the original `dirty` set because it expands internally (same as before).

**Comment by djdanielsson** [comment]
@coderabbitai full code review

File: src/apme_engine/graph/audit_metadata.py [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+            name = patched.get("name")
+            if isinstance(name, str) and var_looks_sensitive(name):
+                patched["name"] = REDACTED
+            val = patched.get("value")
+            if isinstance(val, (dict, list)):
+                patched["value"] = redact_sensitive_structure(val)
+            elif val is not None and (
+                (var_looks_sensitive(str(name)) if name is not None else False) or value_looks_sensitive(val)
+            ):
+                patched["value"] = REDACTED
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/audit_metadata.py:73]: Pass redact_all_scalars=True to redact_sensitive_structure when the variable_set entry name is sensitive, so nested composite values like current: secret are fully redacted. Added regression test. Fixed in e050faa0.

File: tests/test_graph_scanner.py [hunk 20]
```diff
@@ -324,3 +324,26 @@ def test_all_graph_rules_load_without_errors(self) -> None:
...
+    def test_default_load_skips_disabled_by_default_rules(self) -> None:
+        """Disabled-by-default rules are omitted when no rule_id_list is given."""
+        from pathlib import Path
+
+        import apme_engine.validators.native.rules as rules_pkg
+
+        rules_dir = str(Path(rules_pkg.__file__).parent)
+        rule_ids = {r.rule_id for r in load_graph_rules(rules_dir=rules_dir)[0]}
+        assert "R402" not in rule_ids
+        assert "R404" not in rule_ids
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tests/test_graph_scanner.py:337]: Updated test_default_load_skips_disabled_by_default_rules to use native_rules_dir() from graph.scanner, which points at apme_engine.graph.rules. Fixed in e050faa0.

File: tools/generate_rule_catalog.py [hunk 21]
```diff
@@ -120,32 +140,60 @@ def _get_fixable_ids() -> set[str]:
...
+        rf"def\s+test_\w*{re.escape(rid_lower)}\w*|"
+        rf"\bRule\b.*{re.escape(rule_id)})",
+        re.IGNORECASE,
+    )
+    if rule_context.search(text):
+        return False
+    rule_id_assignment = re.compile(
+        rf'["\']?rule_id["\']?\s*[=:]\s*["\']{re.escape(rule_id)}["\']',
+    )
+    return not rule_id_assignment.search(text)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tools/generate_rule_catalog.py]: Replaced generic rule_id fixture matching with explicit # rule-coverage: markers and removed the broad Rule.* pattern that matched RuleMetadata fixtures. Regenerated RULE_CATALOG.md; R402/R404 and tested total now 85/156. Fixed in e050faa0.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/audit_metadata.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+        val: Structure to serialize (list, dict, etc.).
+        rule_id: Rule ID for error logging (optional).
+        key: Metadata key name for error logging (optional).
+
+    Returns:
+        JSON string, or None if serialization fails.
+    """
+    sanitized = sanitize_audit_metadata_value(key, val) if key in AUDIT_JSON_METADATA_KEYS else val
+    try:
+        return json.dumps(sanitized, default=str, sort_keys=True)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/audit_metadata.py:125]: Removed default=str from serialize_audit_metadata_value and build_audit_metadata_blob so unsupported values hit the existing failure path and return None/empty string instead of leaking __str__ output. Added regression test with a stringifiable object. Fixed in bad033ef.

File: src/apme_engine/graph/rules/R401_list_all_inbound_src_graph.py [hunk 23]
```diff
@@ -85,7 +86,12 @@ def process(self, graph: ContentGraph, node_id: str) -> GraphRuleResult | None:
                 continue
             src = resolve_field(mo, profile, "src")
             if src:
-                src_list.append(src)
+                if isinstance(src, str):
+                    src_list.append(redact_url_userinfo(src))
+                elif isinstance(src, (dict, list)):
+                    src_list.append(redact_sensitive_structure(src))
+                else:
+                    src_list.append(src)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/rules/R401_list_all_inbound_src_graph.py]: Extended redact_url_userinfo to redact sensitive query parameter values (e.g. password=hunter2) in addition to userinfo. R401 and audit_metadata inbound_src sanitization both use this helper. Added regression tests. Fixed in bad033ef.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/audit_metadata.py [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+            else:
+                sanitized.append(entry)
+        return sanitized
+
+    if key == "variable_set" and isinstance(value, list):
+        sanitized_vars: list[object] = []
+        for entry in value:
+            if not isinstance(entry, dict):
+                sanitized_vars.append(entry)
+                continue
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/audit_metadata.py:62]: Added `_sanitize_list_scalar` so non-dict `variables_used` and `variable_set` entries are checked with `var_looks_sensitive` / `value_looks_sensitive` before persistence. Regression tests cover scalar strings like `db_password=s3cret` and sensitive variable names. Fixed in 851f47d4.

File: src/apme_engine/graph/variable_helpers.py [hunk 25]
```diff
@@ -0,0 +1,316 @@
...
+            out.append(v)
+        elif isinstance(v, dict):
+            _collect_dict_strings(v, out)
+        elif isinstance(v, list):
+            for item in v:
+                if isinstance(item, str):
+                    out.append(item)
+                elif isinstance(item, dict):
+                    _collect_dict_strings(item, out)
+
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/variable_helpers.py]: Added `depth` / `max_depth=32` parameters to `_collect_dict_strings`, mirroring `redact_sensitive_structure` in `sensitivity.py`. Regression test confirms deeply nested dicts (40 levels) stop collecting past the limit. Fixed in 851f47d4.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: tools/generate_rule_catalog.py [hunk 26]
```diff
@@ -230,6 +280,26 @@ def _find_doc(rule_id: str, doc_dir: Path) -> Path | None:
...
+def _normalize_validator(raw: str) -> str:
+    """Normalize a validator name preserving canonical casing.
+
+    Args:
+        raw: Raw validator string from frontmatter.
+
+    Returns:
+        Canonical validator name, defaulting to ``Native``.
+    """
+    return _VALIDATOR_NAMES.get(raw.lower().strip(), raw.capitalize()) if raw else "Native"
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply tools/generate_rule_catalog.py:323]: Updated `_normalize_validator` to return `"Native"` for any value outside `_VALIDATOR_NAMES`, matching the documented fallback and ensuring unknown frontmatter values still appear in By Validator sections. Test expectation updated accordingly. Fixed in ef585132.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/audit_metadata.py [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+            name_is_sensitive = isinstance(name, str) and var_looks_sensitive(name)
+            if isinstance(val, (dict, list)):
+                patched["value"] = redact_sensitive_structure(
+                    val,
+                    redact_all_scalars=name_is_sensitive,
+                )
+            elif val is not None and (name_is_sensitive or value_looks_sensitive(val)):
+                patched["value"] = REDACTED
+            sanitized_vars.append(patched)
+        return sanitized_vars
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/audit_metadata.py:75]: Added an explicit `isinstance(val, str)` guard before `value_looks_sensitive(val)` in the variable_set scalar branch, matching the defensive pattern used elsewhere in this module. Non-string values now pass through to the existing serialization drop path without invoking the sensitivity check. Fixed in 13deca63.

**Comment by djdanielsson** [comment]
@coderabbitai full review

File: src/apme_engine/graph/sensitivity.py [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+        "credential",
+        "credentials",
+        "cred",
+        "authorization",
+        "private_key",
+        "ssh_key",
+        "access_key",
+        "client_key",
+    }
+)
```
[the remark this answers is not in the document]
Comment (djdanielsson) [reply src/apme_engine/graph/sensitivity.py:28]: Added `passphrase` to `_SENSITIVE_WORDS` so `vault_passphrase` and `ssh_passphrase` are flagged by `var_looks_sensitive`, and added regression tests in `TestVarLooksSensitive::test_passphrase_variants`. Fixed in bb969a1c.

**Comment by djdanielsson** [comment]
@coderabbitai full review

**Comment by djdanielsson** [comment]
@cidrblock friendly reminder about this one
